flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveScheduler.java

```diff
@@ -248,9 +250,24 @@ public AdaptiveScheduler(
                         declarativeSlotPool::reserveFreeSlot,
                         declarativeSlotPool::freeReservedSlot);
 
-        for (JobVertex vertex : jobGraph.getVertices()) {
-            if (vertex.getParallelism() == ExecutionConfig.PARALLELISM_DEFAULT) {
-                vertex.setParallelism(1);
+        if (configuration.get(JobManagerOptions.SCHEDULER_MODE)
+                == SchedulerExecutionMode.REACTIVE) {
+            LOG.info("Modifying job parallelism for running in reactive mode.");
+            for (JobVertex vertex : jobGraph.getVertices()) {
+                if (vertex.getMaxParallelism() == JobVertex.MAX_PARALLELISM_DEFAULT) {
+                    vertex.setParallelism(Transformation.UPPER_BOUND_MAX_PARALLELISM);
+                    vertex.setMaxParallelism(Transformation.UPPER_BOUND_MAX_PARALLELISM);
+                } else {
+                    vertex.setParallelism(vertex.getMaxParallelism());
+                }
+            }
+        } else {
+            // non-reactive mode (test execution with adaptive scheduler): ensure parallelism is set
+            // for all vertices.
+            for (JobVertex vertex : jobGraph.getVertices()) {
+                if (vertex.getParallelism() == ExecutionConfig.PARALLELISM_DEFAULT) {
+                    vertex.setParallelism(1);
+                }
             }
```

Comment:
I think it should not be the `AdaptiveScheduler's` responsibility to set up the job correctly. If we do it, then we make the `AdaptiveScheduler` aware of the `ReactiveMode` which it does not need to know. The reactive mode is a user of the `AdaptiveScheduler` and thus needs to know about the `AdaptiveScheduler`. The other direction is not necessary.

Reply:
Moreover, I would suggest to required that the `JobGraph` is already properly configured when it arrives at the `AdaptiveScheduler`.

Reply:
This is just an explanation for my thinking, not a counter-argument: The reason why I've put this here was because I assume that reactive mode is the only mode currently supported by the adaptive scheduler, and putting it here is the simplest approach (KISS-style). Once the requirements change, we can move this to an outside location.

Reply:
KISS is not a justification to violate SOLID.